lib/trino-parquet/src/main/java/io/trino/parquet/DataPage.java

```diff
@@ -17,11 +17,21 @@
         extends Page
 {
     protected final int valueCount;
+    private final long firstRowIndex;
 
-    public DataPage(int uncompressedSize, int valueCount)
+    public DataPage(int uncompressedSize, int valueCount, long firstRowIndex)
     {
         super(uncompressedSize);
         this.valueCount = valueCount;
+        this.firstRowIndex = firstRowIndex;
+    }
+
+    /**
+     * @return the index of the first row index in this page or -1 if unset.
```

Comment:
This value is the index of the first row in this page across all the rows for the column in the row group. It's read from the column index. It's used to help synchronize the reading across columns (in combination with the offset information in the column index). It's set to -1 in the PageReader.getFirstRowIndex() call if there's no offset index information (called by PageReader.readPage()).

Instead of having an unset value, it's probably cleaner to use an Optional<Long>.